src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -95651,6 +95648,241 @@ class Box<T>
+@"#nullable enable
+using System;
+
+public class Program
+{
+    static void Main()
+    {
+        string? x = ""42"";
+        string? y = ""43"";
+        if (x?.Length == 2 && y?.Length == 2)
+        {
+            Console.WriteLine(x.Length);
+            Console.WriteLine(y.Length);
+        }
+        else
+        {
+            Console.WriteLine(x.Length); // 1
+            Console.WriteLine(y.Length); // 2
+        }
+    }
+}";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics(
+                    // (17,31): warning CS8602: Dereference of a possibly null reference.
+                    //             Console.WriteLine(x.Length); // 1
+                    Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "x").WithLocation(17, 31),
+                    // (18,31): warning CS8602: Dereference of a possibly null reference.
+                    //             Console.WriteLine(y.Length); // 2
+                    Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "y").WithLocation(18, 31));
+        }
+
+        [Fact, WorkItem(34942, "https://github.com/dotnet/roslyn/issues/34942")]
+        public void ConditionalAccess_05()
+        {
+            var source =
+@"#nullable enable
+using System;
+
+public class Program
+{
+    static void Main()
+    {
+        string? x = ""42"";
+        string? y = ""43"";
+        if (x?.Length != 2 || y?.Length != 2)
+        {
+            Console.WriteLine(x.Length); // 1
+            Console.WriteLine(y.Length); // 2
+        }
+        else
+        {
+            Console.WriteLine(x.Length);
+            Console.WriteLine(y.Length);
+        }
+    }
+}";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics(
+                    // (12,31): warning CS8602: Dereference of a possibly null reference.
+                    //             Console.WriteLine(x.Length); // 1
+                    Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "x").WithLocation(12, 31),
+                    // (13,31): warning CS8602: Dereference of a possibly null reference.
+                    //             Console.WriteLine(y.Length); // 2
+                    Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "y").WithLocation(13, 31));
+        }
+
+        [Fact, WorkItem(34942, "https://github.com/dotnet/roslyn/issues/34942")]
+        public void ConditionalAccess_06()
+        {
+            var source =
+@"#nullable enable
+using System;
+
+public class Program
+{
+    static void Main()
+    {
+        string? x = ""42"";
+        string? y = ""43"";
+        if (x?.Length == 2 ? y?.Length == 2 : y?.Length == 3)
+        {
+            Console.WriteLine(x.Length); // 1
+            Console.WriteLine(y.Length);
+        }
+        else
+        {
+            Console.WriteLine(x.Length); // 2
+            Console.WriteLine(y.Length); // 3
+        }
+    }
+}";
+            // The first warning produced for y.Length is unexpected.
```

Comment:
>The first warning produced for y.Length is unexpected. [](start = 15, length = 54)

Actually, I'm surprised by warning 3 too. Why do we get that?

Reply:
In warning 3, we only know that one of `y?.Length != 2` or `y?.Length != 3` is true. So basically we haven't ruled out null.

I believe the issue is that our conditional expression analysis needs to be more complex to handle cases like these. More detail in the tracked issue #36096